torch/csrc/jit/script/sugared_value.h

```diff
@@ -157,12 +167,50 @@ struct TORCH_API BuiltinModule : public SugaredValue {
   c10::optional<int64_t> version;
 };
 
+// TODO right now this is doing double duty as a type and an instance of that
+// type, need to separate
+struct TORCH_API UserTypeValue : public SugaredValue {
```

Comment:
Values are first class but all the attribute logic is being handled in a SugaredValue. This isn't going to work in many cases. For instance, once we conditionally assign a UserType, then it will no longer be a SugaredValue.

It makes sense for UserTypeValue to handle the special logic for `__init__` assignment but otherwise the logic for normal assignment and normal attribute lookup should be defined in SimpleValue. UserTypeValue is probably a subclass of SimpleValue.